### PR TITLE
RATIS-1551. Suppport listener in setConfiguration

### PR DESCRIPTION
**Pull request by codings-dan**
## What changes were proposed in this pull request?

Sub task to support listener:  Suppport listener in setConfiguration
## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-1551

## How was this patch tested?
UT

### CODE REVIEW

**Comment by szetszwo** [comment]
@codings-dan , It seems that we don't need a different LogAppender for Listeners since LogAppender just send raft log to the other side.  So how about we keep using FollowerInfo for Listeners?  We may change the interface name to PeerInfo later on.

In LeaderStateImpl, there are senders and voterLists.
- senders could include both Followers and Listeners.
- voterLists only includes Followers.

What do you think?

**Comment by szetszwo** [comment]
> ...  add corresponding methods to distinguish Listener and follower.

@codings-dan , maybe.  The role info (Listener or Follower) should comes from RaftConfiguration.  We are currently passing a RaftConfiguration to divdeFollowers(..) in LeaderStateImpl in order to get voterLists.  I wonder if it is already good enough, i.e. we can use voterLists to distinguish Listener and Follower.  We use voterLists for calculating majority; see getMajorityMin(..).

**Review by szetszwo — CHANGES_REQUESTED** [review]
@codings-dan , thanks for working on this.

Just found that if we pass RaftPeerRole as a parameter (instead of adding Listener methods), then the number of cases will be reduced; see https://issues.apache.org/jira/secure/attachment/13041243/622_review.patch

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 1]
```diff
@@ -235,6 +235,7 @@ boolean removeAll(Collection<LogAppender> c) {
   private volatile ConfigurationStagingState stagingState;
   private List<List<RaftPeerId>> voterLists;
   private final Map<RaftPeerId, FollowerInfo> peerIdFollowerInfoMap = new ConcurrentHashMap<>();
+  private final Map<RaftPeerId, FollowerInfo> peerIdListenerInfoMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (codings-dan) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:238]: Yean,  adding a data structure is really hard to keep it correct.

**Comment by codings-dan** [comment]
@szetszwo Thanks for helping change the code, the introduction of this variable really reduces a lot of redundant code!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by codings-dan** [comment]
@szetszwo Now that the `SetConfiguration` part has been developed, can we reuse the code of `FollowerInfo` and `FollowerState` in the next step to support Listener, so that the related code of listener is completed.

**Pull request merged**